cmd/skuba/cluster/init.go

```diff
@@ -32,6 +35,22 @@ type initOptions struct {
 	StrictCapDefaults bool
 }
 
+// Checks if clustername provided is correct
+func isValidClusterName(clustername string) bool {
+	if len(clustername) == 0 {
+		return false
+	}
+	// NOT a letter, a number, underscore/dash
+	reg, err := regexp.Compile("[^a-zA-Z0-9_-]+")
```

Comment:
Not sure I agree with this change. This was working before:

```
$ skuba cluster init --control-plane 192.168.122.100 ⛄
** This is an UNTAGGED version and NOT intended for production usage. **
[init] configuration files written to /home/colleen/clusters/⛄
$ cd ./⛄
$ skuba node bootstrap --user devuser --sudo --target 192.168.122.100 master
** This is an UNTAGGED version and NOT intended for production usage. **
W0521 15:39:47.633782    7929 ssh.go:311] 
The authenticity of host '192.168.122.100:22' can't be established.
ECDSA key fingerprint is 4d:0e:01:f0:bd:af:a0:5b:ad:4a:98:61:0e:8a:89:e0.
I0521 15:39:47.633860    7929 ssh.go:312] accepting SSH key for "192.168.122.100:22"
I0521 15:39:47.633871    7929 ssh.go:313] adding fingerprint for "192.168.122.100:22" to "known_hosts"
[bootstrap] updating init configuration with target information
[bootstrap] writing init configuration for node
[bootstrap] applying init configuration to node
E0521 15:42:53.636361    7929 ssh.go:195] W0521 22:42:53.382455    2328 removeetcdmember.go:79] [reset] No kubeadm config, using etcd pod spec to get data directory
E0521 15:42:55.662885    7929 ssh.go:195] W0521 22:42:55.409223    2328 cleanupnode.go:81] [reset] Failed to remove containers: output: time="2020-05-21T22:42:55Z" level=fatal msg="failed to connect: failed to connect, make sure you are running as root and the runtime has been started: context deadline exceeded"
E0521 15:42:55.662914    7929 ssh.go:195] , error: exit status 1
E0521 15:42:56.845857    7929 ssh.go:195] No files found for firewalld.service.
E0521 15:43:00.836760    7929 ssh.go:195] Created symlink /etc/systemd/system/multi-user.target.wants/crio.service → /usr/lib/systemd/system/crio.service.
E0521 15:43:02.751170    7929 ssh.go:195] Created symlink /etc/systemd/system/multi-user.target.wants/kubelet.service → /usr/lib/systemd/system/kubelet.service.
E0521 15:43:03.165058    7929 ssh.go:195] W0521 22:43:02.911409    3877 configset.go:202] WARNING: kubeadm cannot validate component configs for API groups [kubelet.config.k8s.io kubeproxy.config.k8s.io]
E0521 15:43:03.201876    7929 ssh.go:195]       [WARNING Hostname]: hostname "master" could not be reached
E0521 15:43:03.201906    7929 ssh.go:195]       [WARNING Hostname]: hostname "master": lookup master on 192.168.122.1:53: no such host
E0521 15:46:16.262865    7929 ssh.go:195] W0521 22:46:16.007950    3877 manifests.go:225] the default kube-apiserver authorization-mode is "Node,RBAC"; using "Node,RBAC"
E0521 15:46:16.263964    7929 ssh.go:195] W0521 22:46:16.009205    3877 manifests.go:225] the default kube-apiserver authorization-mode is "Node,RBAC"; using "Node,RBAC"
E0521 15:46:46.065721    7929 ssh.go:195] Created symlink /etc/systemd/system/timers.target.wants/skuba-update.timer → /usr/lib/systemd/system/skuba-update.timer.
[bootstrap] successfully bootstrapped core components on node "192.168.122.100" with Kubernetes: "1.18.2"
[bootstrap] downloading secrets from bootstrapped node "192.168.122.100"
[bootstrap] deploying core add-ons on node "192.168.122.100"
[bootstrap] successfully bootstrapped core add-ons on node "192.168.122.100"
$ kubectl get node
NAME     STATUS   ROLES    AGE     VERSION
master   Ready    master   5m35s   v1.18.2
```

Seems like an unnecessary restriction that's unrelated to the bug that was reported.

Reply:
That's interesting. Neither Kind nor Kops allow those names. Kubeadm is not testing those use cases, so I suppose everything is allowed from kubeadm perspective.
I thought we had to restrict the cluster names (for gangway), similarly to those tools above. I took the more lenient approach than the strict dns name validation checking.
You're right on the fact this isn't ready. Either I should be stricter, or looser.

Reply:
Stricter here could mean follow the dns rfc, or do like docker: 
https://github.com/moby/moby/blob/master/daemon/names/names.go#L6

It seems crio doesn't have that restriction, runC doesn't validate those names, and kubeadm doesn't test those, so that's why it works. The safeguards in kind are implemented because it uses docker/moby behind the scenes, which _is_ restricted. 

I will relax this.